ext/mbstring/libmbfl/filters/mbfilter_cp936.c

```diff
@@ -291,37 +291,41 @@ static size_t mb_cp936_to_wchar(unsigned char **in, size_t *in_len, uint32_t *bu
 			}
 
 			unsigned char c2 = *p++;
+			if (c2 < 0x40 || c2 == 0x7F || c2 == 0xFF) {
+				*out++ = MBFL_BAD_INPUT;
+				continue;
+			}
 
-			if (((c >= 0xAA && c <= 0xAF) || (c >= 0xF8 && c <= 0xFE)) && (c2 >= 0xA1 && c2 <= 0xFE)) {
+			if (((c >= 0xAA && c <= 0xAF) || (c >= 0xF8 && c <= 0xFE)) && c2 >= 0xA1) {
 				/* UDA part 1, 2: U+E000-U+E4C5 */
 				*out++ = 94*(c >= 0xF8 ? c - 0xF2 : c - 0xAA) + (c2 - 0xA1) + 0xE000;
-			} else if (c >= 0xA1 && c <= 0xA7 && c2 >= 0x40 && c2 < 0xA1 && c2 != 0x7F) {
+			} else if (c >= 0xA1 && c <= 0xA7 && c2 < 0xA1) {
 				/* UDA part 3: U+E4C6-U+E765*/
 				*out++ = 96*(c - 0xA1) + c2 - (c2 >= 0x80 ? 0x41 : 0x40) + 0xE4C6;
 			} else {
-				unsigned int w = (c << 8) | c2;
-
-				if ((w >= 0xA2AB && w <= 0xA9FE) || (w >= 0xD7FA && w <= 0xD7FE) || (w >= 0xFE50 && w <= 0xFEA0)) {
-					for (int k = 0; k < mbfl_cp936_pua_tbl_max; k++) {
-						if (w >= mbfl_cp936_pua_tbl[k][2] && w <= mbfl_cp936_pua_tbl[k][2] + mbfl_cp936_pua_tbl[k][1] - mbfl_cp936_pua_tbl[k][0]) {
-							*out++ = w -  mbfl_cp936_pua_tbl[k][2] + mbfl_cp936_pua_tbl[k][0];
-							goto next_iteration;
+				unsigned int w = (c - 0x81)*192 + c2 - 0x40;
```

Comment:
Can you add a comment to this line, I have no clue what this is doing and why.

Reply:
OK, adding comment now.

Reply:
Basically, we are converting the 2 byte values for a 2-byte character into a table lookup index.

The high bits of the lookup index come from the 1st byte, the low bits come from the 2nd byte.

2nd byte values start from `0x40`, so we subtract off `0x40` so that the lookup indices start from 0. Likewise, 1st byte values start from `0x81`, so...

And then, each different value for the 1st byte correspond to a block of 192 values in the lookup table, so that's why we multiply by 192.

Reply:
Comment added.